src/app/clusters/network-commissioning/network-commissioning.cpp

```diff
@@ -546,18 +548,63 @@ void Instance::OnFinished(Status status, CharSpan debugText, ThreadScanResponseI
         err = writer->StartContainer(TLV::ContextTag(to_underlying(Commands::ScanNetworksResponse::Fields::kThreadScanResults)),
                                      TLV::TLVType::kTLVType_Array, listContainerType));
 
-    for (; networks != nullptr && networks->Next(scanResponse) && networksEncoded < kMaxNetworksInScanResponse; networksEncoded++)
+    VerifyOrExit(
+        scanResponseArray.Alloc(networks->Count() > kMaxNetworksInScanResponse ? kMaxNetworksInScanResponse : networks->Count()),
+        err = CHIP_ERROR_NO_MEMORY);
+    for (; networks != nullptr && networks->Next(scanResponse);)
+    {
+        if ((scanResponseArrayLength == kMaxNetworksInScanResponse) &&
+            (scanResponseArray[scanResponseArrayLength - 1].rssi > scanResponse.rssi))
+        {
+            continue;
+        }
+
+        bool isDuplicated = false;
+
+        for (size_t i = 0; i < scanResponseArrayLength; i++)
+        {
+            if ((scanResponseArray[i].panId == scanResponse.panId) &&
+                (scanResponseArray[i].extendedPanId == scanResponse.extendedPanId))
+            {
+                isDuplicated = true;
+                if (scanResponseArray[i].rssi < scanResponse.rssi)
+                {
+                    scanResponseArray[i] = scanResponse;
+                    Sorting::InsertionSort(
```

Comment:
Note that given that you sort elements each time you add a new one, you could here do just:
```
scanResponseArray[i] = scanResponseArray[--scanResponseArrayLength]
```
to remove the duplicated element (to be re-added below). But this is a minor optimization :)

Reply:
Thank you for your suggestion! I understood your suggestion as below, right? However, if I change it like this, the sort occurs twice (once when deleted, once when added). Please let me know if I misunderstood.

```diff
             if ((scanResponseArray[i].panId == scanResponse.panId) &&
                 (scanResponseArray[i].extendedPanId == scanResponse.extendedPanId))
             {
-                isDuplicated = true;
                 if (scanResponseArray[i].rssi < scanResponse.rssi)
                 {
-                    scanResponseArray[i] = scanResponse;
+                    scanResponseArray[i] = scanResponseArray[--scanResponseArrayLength];
                     Sorting::InsertionSort(
-                        scanResponseArray.Get(), scanResponseArrayLength,
+                        scanResponseArray.Get(), i + 1,
                         [](const ThreadScanResponse & a, const ThreadScanResponse & b) -> bool { return a.rssi > b.rssi; });
                 }
+                else
+                {
+                    isDuplicated = true;
+                }
                 break;
             }
         }
```

Reply:
Yeah, but the sort on deletion (here) is not needed, right?

Reply:
Right. Because of the other comments, I thought you wanted the sort to remain when deleting. I will fix it. Thank you 👍